server/controllers/location.ts

```diff
@@ -0,0 +1,165 @@
+        await this.databaseSdk.find('location', request).then(response => {
+            response = _.map(response['docs'], (doc) => requestBody.locationType === 'state' ? _.omit(doc, ['_id', '_rev', 'data']) : doc.data);
+            let resObj = {
+                response: requestBody.locationType === 'district' ? response[0] : response
+            };
+            logger.info(`ReqId =  ${req.headers['X-msgid']}: got data from db`);
+            return res.send(Response.success('api.location.read', resObj, req));
+        }).catch(err => {
+            logger.error(
+                `ReqId = "${req.headers[
+                'X-msgid'
+                ]}": Received error while searching in location database and err.message: ${err.message} ${err}`
+            );
+            if (err.status === 404) {
+                res.status(404);
+                return res.send(Response.error('api.location.read', 404));
+            } else {
+                let status = err.status || 500;
+                res.status(status);
+                return res.send(Response.error('api.location.read', status));
+            }
+        });
+    }
+    async proxyToAPI(req, res, next) {
+
+        let requestObj = {
+            type: _.get(req.body, 'request.filters.type'),
+            parentId: _.get(req.body, 'request.filters.parentId')
+        }
+        const config = {
+            headers: {
+                authorization: `Bearer ${process.env.APP_BASE_URL_TOKEN}`,
+                'content-type': 'application/json',
+            },
+        };
+        const filter = _.isEmpty(requestObj.parentId)
+            ? { filters: { type: requestObj.type } }
+            : { filters: { type: requestObj.type, parentId: requestObj.parentId } };
+
+        const requestParams = {
+            request: filter,
+        };
+        try {
+            logger.debug(`ReqId =  ${req.headers["X-msgid"]}}: getting location data from online`);
+            let responseData = await HTTPService.post(
+                `${process.env.APP_BASE_URL}/api/data/v1/location/search`,
+                requestParams,
+                config
+            ).toPromise();
+
+            let resObj = {
+                response: _.get(responseData.data, 'result.response')
+            }
+
+            logger.debug(`ReqId =  ${req.headers["X-msgid"]}}: fetchLocationFromOffline method is calling `);
+            await this.insertOnlineDataInDB(resObj, requestObj, req.headers['X-msgid']);
+            resObj.response = _.map(resObj.response , data => _.has(data, 'data') ? _.omit(data, 'data') : data);
+            return res.send(Response.success('api.location.read', resObj, req));
+        } catch (err) {
+            logger.error(`Error Received while getting data from Online ${_.get(err, 'response')}`)
+            next();
+        }
+    }
+
+    async insertOnlineDataInDB(onlineData, requestObj, msgId) {
```

Comment:
locationType === “state”
 if(got data) 
onlineStates  = []  ids = _.map()  
[] , [saf] , [ids] = find(ids)

bulkInsert = []
bulkUpdate = []

For (state of onlineStates ) {  if(state.id includes ids) {
_.find(onlineStates, {id: state.id})
// add data = state.data || []
bulkUpdate.push(buldadd data ) 
} else 
State.data = [] bulkInsert.push(state)  } if(bulkInsert.length > 0)
bulkinsert 
if(bulkupdate)
bulkUpdate  else  skip 

return   locationType === “district” and parentId  if(got Data)  state = db.get(locations, parentId)
 state.data = data

db.update(state) 
return data else  skip 

return;